### ISSUE TITLE
Cannot open FTPS encrypted data connection through HTTP proxy

### ISSUE THREAD

**Issue by starless72** [bagder/curl#278]
When trying to transfer a file via FTPS through a HTTP proxy (Blue Coat ProxySG), curl fails when trying to open the encrypted DATA connection returning the following error:

`curl: (35) error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number`

This surely happens with recent curl versions (7.39 and 7.40), on Red Hat 6.3.
The transfer works fine instead with previous versions, like 7.19.7, 7.25.0, 7.26.0 and 7.27.0.

Here https://gist.github.com/starless72/64de772a1397bf0e2da3 you can see the log files obtained with the -v option both from 7.40 (not working) and 7.19 (working), from the same Red Hat 6.3 machine over the same proxy.
The logs also contain the output of curl -V and the command line used:

`curl -O -v -k --ftp-ssl --disable-epsv --proxy <proxyIP>:3128 --proxytunnel ftp://<ftpUser>:<ftpPassword>@<serverIP>//testpub.xml`

Also note that when disabling data channel encryption with the --ftp-ssl-control option then the connection occurs but a different problem happens during the file transfer, see issue #277.

**Comment by starless72** [comment]
The problem is still present in a more recent snapshot.
I downloaded http://curl.haxx.se/snapshots/curl-7.42.0-20150331.tar.gz, built it with default configure, make, make install on another RHEL 6.3 machine, then copied the /usr/local/* curl-related content on the target machine which is behind the proxy, and launched the same test command using /usr/local/bin/curl, but I get the exact same result, the downloaded file is not complete.
I hope the process I used is OK. I do not have complete freedom of what I can do on the target machine. Here is what I get with curl -V after installing the daily snapshot this way:

```
# /usr/local/bin/curl -V
curl 7.42.0-20150331 (x86_64-unknown-linux-gnu) libcurl/7.42.0-20150331 OpenSSL/1.0.0 zlib/1.2.3 libidn/1.18
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: IDN IPv6 Largefile NTLM NTLM_WB SSL libz UnixSockets
```

**Comment by bagder** [comment]
How is this different than #277 or perhaps when does it work and when doesn't it work?

**Comment by starless72** [comment]
This issue is about establishing an **encrypted data connection**, i.e. when using `--ftp-ssl` then the data connection cannot be started at all.
#277 is instead about transferring a file over an **unencrypted** data channel, when using `--ftp-ssl-control`.

It *never* works :) with current curl version. With encrypted data the transfer cannot be started. With unencrypted data the transfer is started but the file is truncated.

**Comment by bagder** [comment]
I can now repeat this problem. I'll investigate.

**Comment by bagder** [comment]
Here's my suggested fix. Too close to the 7.43.0 release time to get merged before the release but I intend to merge immediately after unless someone finds problems with it.

~~~
From c72e9cdac6ca8950ca4b024f4439d1b72fce9bf0 Mon Sep 17 00:00:00 2001
From: Daniel Stenberg <daniel@haxx.se>
Date: Wed, 17 Jun 2015 00:30:06 +0200
Subject: [PATCH] FTP: do the HTTP CONNECT for data connection blocking

** WORK-AROUND **

The introduced non-blocking general behaviour for Curl_proxyCONNECT()
didn't work for the data connection establishment unless it was very
fast. The newly introduced function argument makes it operate in a more
blocking manner, more like it used to work in the past. This blocking
approach is only used when the FTP data connecting through HTTP proxy.

Blocking like this is bad. A better fix would make it work more
asynchronously.

Bug: https://github.com/bagder/curl/issues/278
---
 lib/ftp.c        |  4 ++--
 lib/http_proxy.c | 22 ++++++++++++++--------
 lib/http_proxy.h |  5 +++--
 3 files changed, 19 insertions(+), 12 deletions(-)

diff --git a/lib/ftp.c b/lib/ftp.c
index de628a2..fade092 100644
--- a/lib/ftp.c
+++ b/lib/ftp.c
@@ -1885,11 +1885,11 @@ static CURLcode proxy_magic(struct connectdata *conn,
     struct HTTP http_proxy;
     struct FTP *ftp_save = data->req.protop;
     memset(&http_proxy, 0, sizeof(http_proxy));
     data->req.protop = &http_proxy;
 
-    result = Curl_proxyCONNECT(conn, SECONDARYSOCKET, newhost, newport);
+    result = Curl_proxyCONNECT(conn, SECONDARYSOCKET, newhost, newport, TRUE);
 
     data->req.protop = ftp_save;
 
     if(result)
       return result;
@@ -3643,11 +3643,11 @@ static CURLcode ftp_do_more(struct connectdata *conn, int *completep)
   /* if the second connection isn't done yet, wait for it */
   if(!conn->bits.tcpconnect[SECONDARYSOCKET]) {
     if(conn->tunnel_state[SECONDARYSOCKET] == TUNNEL_CONNECT) {
       /* As we're in TUNNEL_CONNECT state now, we know the proxy name and port
          aren't used so we blank their arguments. TODO: make this nicer */
-      result = Curl_proxyCONNECT(conn, SECONDARYSOCKET, NULL, 0);
+      result = Curl_proxyCONNECT(conn, SECONDARYSOCKET, NULL, 0, FALSE);
 
       return result;
     }
 
     result = Curl_is_connected(conn, SECONDARYSOCKET, &connected);
diff --git a/lib/http_proxy.c b/lib/http_proxy.c
index 5ab9915..4373d62 100644
--- a/lib/http_proxy.c
+++ b/lib/http_proxy.c
@@ -66,11 +66,11 @@ CURLcode Curl_proxy_connect(struct connectdata *conn)
     prot_save = conn->data->req.protop;
     memset(&http_proxy, 0, sizeof(http_proxy));
     conn->data->req.protop = &http_proxy;
     connkeep(conn, "HTTP proxy CONNECT");
     result = Curl_proxyCONNECT(conn, FIRSTSOCKET,
-                               conn->host.name, conn->remote_port);
+                               conn->host.name, conn->remote_port, FALSE);
     conn->data->req.protop = prot_save;
     if(CURLE_OK != result)
       return result;
     Curl_safefree(conn->allocptr.proxyuserpwd);
 #else
@@ -83,16 +83,20 @@ CURLcode Curl_proxy_connect(struct connectdata *conn)
 
 /*
  * Curl_proxyCONNECT() requires that we're connected to a HTTP proxy. This
  * function will issue the necessary commands to get a seamless tunnel through
  * this proxy. After that, the socket can be used just as a normal socket.
+ *
+ * 'blocking' set to TRUE means that this function will do the entire CONNECT
+ * + response in a blocking fashion. Should be avoided!
  */
 
 CURLcode Curl_proxyCONNECT(struct connectdata *conn,
                            int sockindex,
                            const char *hostname,
-                           int remote_port)
+                           int remote_port,
+                           bool blocking)
 {
   int subversion=0;
   struct SessionHandle *data=conn->data;
   struct SingleRequest *k = &data->req;
   CURLcode result;
@@ -223,16 +227,18 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
     if(check <= 0) {
       failf(data, "Proxy CONNECT aborted due to timeout");
       return CURLE_RECV_ERROR;
     }
 
-    if(0 == Curl_socket_ready(tunnelsocket, CURL_SOCKET_BAD, 0))
-      /* return so we'll be called again polling-style */
-      return CURLE_OK;
-    else {
-      DEBUGF(infof(data,
-                   "Read response immediately from proxy CONNECT\n"));
+    if(!blocking) {
+      if(0 == Curl_socket_ready(tunnelsocket, CURL_SOCKET_BAD, 0))
+        /* return so we'll be called again polling-style */
+        return CURLE_OK;
+      else {
+        DEBUGF(infof(data,
+               "Read response immediately from proxy CONNECT\n"));
+      }
     }
 
     /* at this point, the tunnel_connecting phase is over. */
 
     { /* READING RESPONSE PHASE */
diff --git a/lib/http_proxy.h b/lib/http_proxy.h
index 2b5e9c9..058a29d 100644
--- a/lib/http_proxy.h
+++ b/lib/http_proxy.h
@@ -5,11 +5,11 @@
  *  Project                     ___| | | |  _ \| |
  *                             / __| | | | |_) | |
  *                            | (__| |_| |  _ <| |___
  *                             \___|\___/|_| \_\_____|
  *
- * Copyright (C) 1998 - 2014, Daniel Stenberg, <daniel@haxx.se>, et al.
+ * Copyright (C) 1998 - 2015, Daniel Stenberg, <daniel@haxx.se>, et al.
  *
  * This software is licensed as described in the file COPYING, which
  * you should have received as part of this distribution. The terms
  * are also available at http://curl.haxx.se/docs/copyright.html.
  *
@@ -24,11 +24,12 @@
 
 #if !defined(CURL_DISABLE_PROXY) && !defined(CURL_DISABLE_HTTP)
 /* ftp can use this as well */
 CURLcode Curl_proxyCONNECT(struct connectdata *conn,
                            int tunnelsocket,
-                           const char *hostname, int remote_port);
+                           const char *hostname, int remote_port,
+                           bool blocking);
 
 /* Default proxy timeout in milliseconds */
 #define PROXY_TIMEOUT (3600*1000)
 
 CURLcode Curl_proxy_connect(struct connectdata *conn);
-- 
2.1.4



~~~

**Comment by bagder** [comment]
Fix merged in commit b88f980a7437, thanks!
